files/scripts/startup_tsa_tsb.py

```diff
@@ -0,0 +1,156 @@
+    else:
+        return subprocess.check_output(['sonic-cfggen', '-d', '-v', config_name.replace('"', "'"), '-n', ns.replace('"', "'")]).strip()
+
+def get_sub_role(asic_ns):
+    sub_role_config = "DEVICE_METADATA['localhost']['sub_role']"
+    sub_role = (get_sonic_config(asic_ns, sub_role_config)).decode()
+    return sub_role
+
+def get_tsa_config(asic_ns):
+    tsa_config = 'BGP_DEVICE_GLOBAL.STATE.tsa_enabled'
+    tsa_ena = (get_sonic_config(asic_ns, tsa_config)).decode()
+    if asic_ns == "":
+        logger.log_info('CONFIG_DB.{} : {}'.format(tsa_config, tsa_ena))
+    else:
+        logger.log_info('{} - CONFIG_DB.{} : {}'.format(asic_ns, tsa_config, tsa_ena))
+    return tsa_ena
+
+def get_tsa_status(num_asics):
+    if num_asics > 1:
+        counter = 0
+        for asic_id in range(int(asic_num)):
+            asic_ns = 'asic{}'.format(asic_id)
+            sub_role = get_sub_role(asic_ns)
+            if sub_role == 'FrontEnd':
+                tsa_enabled = get_tsa_config(asic_ns)
+                if tsa_enabled == 'false':
+                    counter += 1
+        if counter == int(asic_num):
+            return True;
+    else:
+        tsa_enabled = get_tsa_config("")
+        if tsa_enabled == 'false':
+            return True;
+    return False;
+
+def config_tsa():
+    num_asics = multi_asic.get_num_asics()
+    tsa_ena = get_tsa_status(num_asics)
+    if tsa_ena == True:
+        logger.log_info("Configuring TSA")
+        subprocess.check_output(['TSA']).strip()
+    else:
+        if num_asics > 1:
+            logger.log_info("Either TSA is already configured or switch sub_role is not Frontend - not configuring TSA")
+        else:
+            logger.log_info("Either TSA is already configured - not configuring TSA")
+    return tsa_ena
+
+def config_tsb():
+    logger.log_info("startup_tsa_tsb: Configuring TSB")
+    subprocess.check_output(['TSB']).strip()
+    tsb_issued = True
+    return
+
+def start_tsb_timer(interval):
+    global timer
+    logger.log_info("Starting timer with interval {} seconds to configure TSB".format(interval))
+    timer = Timer(int(interval), config_tsb)
+    timer.start()
+    timer.join()
+    return
+
+def print_usage():
+    logger.log_info("Usage: startup_tsa_tsb.py [options] command")
+    logger.log_info("options:")
+    logger.log_info("  -h | --help       : this help message")
+    logger.log_info("command:")
+    logger.log_info("start     : start the TSA/TSB")
+    logger.log_info("stop      : stop the TSA/TSB")
+    return
+
+def reset_env_variables():
+    logger.log_info("Resetting environment variable")
+    os.environ.pop('STARTED_BY_TSA_TSB_SERVICE')
+    return
+
+def start_tsa_tsb(timer):
+    #Configure TSA if it was not configured already in CONFIG_DB
+    tsa_enabled = config_tsa()
+    if tsa_enabled == True:
+        #Start the timer to configure TSB
+        start_tsb_timer(timer)
+    return
+
+def stop_tsa_tsb():
+    reset_env_variables()
```

Comment:
In case the user entered TSA/TSB command manually, shouldn't we explicitly stop/cancel timer thread which we started earlier in start_tsb_timer() ? 

Here we just reset the env variables only.

Reply:
I will set the daemon flag for the Timer thread so that the thread will exit when the main thread is killed by the service stop

Reply:
@judyjoseph , i tested with prctl and the threads started by the main thread are stopped when the service is stopped and main python script is killed with SIGTERM. So we don't need to explicitly stop the timer.